### DISCUSSION TITLE
Support of fastai based models

### DISCUSSION THREAD

**Discussion by oz-mishli** [chr5tphr/zennit#150] [Q&A] [answered]
Hi, 

I attempted to use zennit with a fastai model and it seemed to work, however I wanted to know if this is something that should be supported  (as fastai is based on PyTorch)? Or should I convert the models and the data to "native" PyTorch before using zennit?

**Comment by chr5tphr** [comment, marked as answer]
Hey Oz,

although I have never used fastai before, it seems like all their layers are based on `torch.nn.Module`, and layers like `fastai.layers.ConvLayer` seem to inherit from `torch.nn.Conv2d`, so they should work fine with Zennit.
There are, however, a few points to keep in mind:

- If your model has (non-native PyTorch) custom layers, by default, their gradient will be used. If the gradient is not the correct/ an applicable/ desirable way to attribute the layer, you need to specify a rule for your layer type. To do this, you can specify your own `Composite` as described [here](https://zennit.readthedocs.io/en/latest/how-to/use-rules-composites-and-canonizers.html#abstract-composites). There is also a new feature coming in the next release, which you can get by installing the dev-version with `pip install 'git+https://github.com/chr5tphr/zennit.git', where you can [specify additional maps to built-in Composites](https://zennit.readthedocs.io/en/latest/how-to/use-rules-composites-and-canonizers.html#cooperative-layermapcomposites).
- If your model uses BatchNorm, you can use the `zennit.canonizers.SequentialMergeBatchNorm`, as shown [here](https://zennit.readthedocs.io/en/latest/how-to/use-rules-composites-and-canonizers.html#canonizers). However, this will only work if your model is Sequential (not necessarily using `torch.nn.Sequential`) and BatchNorms followed a single linear layer.
- If your model has skip-connections, which you would like to weight (for ResNet, the `ResNetCanonizer` does this, but it only works for the torchvision version), you need to use the `zennit.layers.Sum()` layer where the outputs are added. However, you can neglect this in the beginning and see how it works for you.

If you want, you can write the architecture you are using here and I can have a look if there are some potential problems. Did you get reasonable-looking heatmaps?

**Reply by oz-mishli** [reply]
Hi Cristopher,

Thanks for the quick response! I will share few details about my setup:

- I'm using the basic IntegratedGradients (IG) only at this point, without composites etc. This is just my first attempt generating heatmaps with zennit. 
- The model I'm using is ResNet, in the library [tsai](https://timeseriesai.github.io/tsai/); tsai is based on fastai and PyTorch.  

So far, I noticed that it is somewhat slower calculating IG on a tsai model compared to a vanilla PyTorch implementation. However, do note that this experiment was conducted on a 1-layer LSTM and not on ResNet, also on a different dataset, so I'm not sure it's a fair comparison. Would you expect slower computation time for heatmaps if using tsai model rather than a vanilla PyTorch model? 

As for the heatmaps, I'm just getting started. I did get reasonable-looking heatmaps with the LSTM for a simpler dataset, however I'm yet to find what I get with a more complicated dataset with ResNet (calculating the heatmaps takes rather long time).

Cheers,
Oz

**Reply by chr5tphr** [reply]
Hey Oz,

IG only uses the gradient, so it can be used on anything that produces a gradient without problems.
The [tsai-ResNet](https://timeseriesai.github.io/tsai/models.ResNet.html) seems actually easier to use than the torchvision-ResNet, since it implements the skip-connection already as a module.
For this, you can simply use [(given the current dev and future release)](https://zennit.readthedocs.io/en/latest/how-to/use-rules-composites-and-canonizers.html#cooperative-layermapcomposites):
```python
from torch.nn import Linear
from tsai.models.layers import Add
from zennit.composites import EpsilonGammaBox
from zennit.rules import Norm, Pass
from zennit.types import BatchNorm

composite = EpsilonGammaBox(
    low=-3.,
    high=3.,
    layer_map=[
        (BatchNorm, Pass()),  # ignore BatchNorm
        (Add, Norm()),  # Add from tsai-ResNet
    ],
)
```

I expect that IG on a 1-layer LSTM is much faster than on ResNet.
I do not expect there is a big difference in runtime between tsai models and vanilla PyTorch-models, tsai-models seem quite vanilla.

**Reply by oz-mishli** [reply]
Excellent, thank you! I'll proceed as suggested and revert if I encounter any issue.

**Reply by oz-mishli** [reply]
Hi Cristopher,

I used the code you listed above and got the following error. This is weird as I use the latest tsai version, 0.3.1. Can you please help? 

> Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevd.py", line 1483, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "main.py", line 60, in <module>
    exp.generate_maps(model, dsets)
  File "explainer.py", line 14, in generate_maps
    composite = EpsilonGammaBox(
TypeError: __init__() got an unexpected keyword argument 'layer_map'

Process finished with exit code 1

**Reply by chr5tphr** [reply]
Hey Oz,
I still have not been able to finish the release with the cooperative layer maps, so you need to install the git version for this to work, e.g. using `pip install 'git+https://github.com/chr5tphr/zennit.git'`
